### PR TITLE
CHIME telescope added to p(z|DM)

### PR DESCRIPTION
**Pull request by lluism**
it simply adds to disk the CHIME_pzdm.npz file created by X notebook, and then loads it for the calculations. 

Also visibility improvements in low-z cases plots (log scales, etc)

### CODE REVIEW

File: frb/dm/prob_dmz.py [hunk 1]
```diff
@@ -176,6 +176,32 @@ def grab_repo_grid():
...
+        dict: Numpy dict from the npz save file
+    """
+
+    # File
+    CHIME_PDM_z_grid_file = os.path.join(
+        resource_filename('frb', 'data'), 'DM',
+        'CHIME_pzdm.npz')
+
+    # assert the file exists
+    if not os.path.isfile(CHIME_PDM_z_grid_file):
```
Comment (profxj) [inline frb/dm/prob_dmz.py:194]: it looks like you have added the file into the Repo (which I guess is fine).

so, can this happen?

File: frb/dm/prob_dmz.py [hunk 2]
```diff
@@ -176,6 +176,32 @@ def grab_repo_grid():
...
+    """
+    Grab the CHIME grid from the Repository
+
+    Returns:
+        dict: Numpy dict from the npz save file
+    """
+
+    # File
+    CHIME_PDM_z_grid_file = os.path.join(
+        resource_filename('frb', 'data'), 'DM',
```
Comment (profxj) [inline frb/dm/prob_dmz.py:190]: `resource_filename` is deprecated

I now use:

`from importlib.resources import files` or something like that

File: frb/galaxies/mag_dm.py [hunk 3]
```diff
@@ -102,20 +109,28 @@ def r_vs_dm_figure(z_min, z_max, z, PzDM, outfile='fig_r_vs_z.png',
...
 
-    ax.set_xlabel(r'$z$')
-    ax.set_ylabel(r'$m_r$')
+    ax.set_xlabel(r'Redshift ($z$)')
+    ax.set_ylabel(r'Apparent Magnitude ($m_r$)')
     ax.set_xlim(xmnx)
     ax.set_ylim(ymnx)
 
-    ax.xaxis.set_major_locator(plt.MultipleLocator(0.5))
+    if z_max <= 0.5:
```
Comment (profxj) [inline frb/galaxies/mag_dm.py:123]: maybe make the log scaling an option instead?

File: frb/scripts/pz_dm.py [hunk 4]
```diff
@@ -16,6 +16,8 @@ def parser(options=None):
                         help="Confidence limits for the z estimate [default is a 95 percent c.l., (2.5,97.5)]")
     parser.add_argument("--magdm_plot", default=False, action='store_true', 
                         help="Plot the host redshift range given DM on the magnitude vs redshift evolution")
+    parser.add_argument("--perfect_telescope", default=False, action='store_true', help="Use a perfect telescope model for the DM-z grid instead of CHIME")
```
Comment (profxj) [inline frb/scripts/pz_dm.py:19]: let's change this to be `--telescope`
with "Ideal" or something like that as the default
and `CHIME`, `DSA`, `ASKAP/ICS`, `ASKAP/CRACO` etc.
as options

File: frb/scripts/pz_dm.py [hunk 5]
```diff
@@ -52,14 +54,26 @@ def main(pargs):
     # Redshift estimates
 
     # Load
-    sdict = prob_dmz.grab_repo_grid()
-    PDM_z = sdict['PDM_z']
+    # consider CHIME telescope
+    sdict = prob_dmz.grab_chime_repo_grid()
```
Comment (profxj) [inline frb/scripts/pz_dm.py:58]: this needs to be an `if, elif` with all the telescope options

File: frb/scripts/pz_dm.py [hunk 6]
```diff
@@ -52,14 +54,26 @@ def main(pargs):
...
     z = sdict['z']
     DM = sdict['DM']
-
-    # Do it
     iDM = np.argmin(np.abs(DM - DM_cosmic))
-    PzDM = PDM_z[iDM, :] / np.sum(PDM_z[iDM, :])
+    PzDM = PDM_z[:,iDM] / np.sum(PDM_z[:,iDM])
+
+    # perfect telescope case
+    if pargs.perfect_telescope:
```
Comment (profxj) [inline frb/scripts/pz_dm.py:66]: see above

File: frb/scripts/pz_dm.py [hunk 7]
```diff
@@ -81,12 +95,17 @@ def main(pargs):
     print(f"The redshift range for your confidence interval {pargs.cl} is:")
     print(f"z = [{z_min:.3f}, {z_max:.3f}]")
     print("")
-    print("WARNING: This all assumes a perfect telescope and a model of the scatter in DM_cosmic (Macqurt+2020)")
+    if pargs.perfect_telescope:
```
Comment (profxj) [inline frb/scripts/pz_dm.py:98]: again

**Review by profxj — CHANGES_REQUESTED** [review]
2 additional requests:

1. Add docs in the https://github.com/FRBs/FRB/blob/main/docs/scripts.rst file
2. Add new tests in https://github.com/FRBs/FRB/blob/main/frb/tests/test_scripts.py

thanks!

File: frb/dm/prob_dmz.py (hunk 1 above)
Comment (lluism) [reply frb/dm/prob_dmz.py:194]: solved

File: frb/dm/prob_dmz.py (hunk 2 above)
Comment (lluism) [reply frb/dm/prob_dmz.py:190]: solved

File: frb/galaxies/mag_dm.py (hunk 3 above)
Comment (lluism) [reply frb/galaxies/mag_dm.py:123]: done

File: frb/scripts/pz_dm.py (hunk 4 above)
Comment (lluism) [reply frb/scripts/pz_dm.py:19]: done

File: frb/scripts/pz_dm.py (hunk 5 above)
Comment (lluism) [reply frb/scripts/pz_dm.py:58]: done

File: frb/scripts/pz_dm.py (hunk 6 above)
Comment (lluism) [reply frb/scripts/pz_dm.py:66]: done

File: frb/scripts/pz_dm.py (hunk 7 above)
Comment (lluism) [reply frb/scripts/pz_dm.py:98]: solved

**Comment by lluism** [comment]
addressed comments and added extra telescopes for which we got grids from the zdm fellas

**Review by profxj — CHANGES_REQUESTED** [review]
a few minor requests

thx!

File: docs/scripts.rst [hunk 8]
```diff
@@ -129,6 +133,16 @@ Here is the usage::
     --cl CL               Confidence limits for the z estimate [default is a 95
                             percent c.l., (2.5,97.5)]
 
+    --magdm_plot          Plot the host redshift range given DM on the magnitude
+                            vs redshift evolution. Default=False.   
+
+    --fig_title FIG_TITLE  title for the figure; e.g., FRBXXXXX
+
+    --telescope TELESCOPE  telescope model for the DM-z grid: CHIME, DSA, Parkes,
+                            FAST, CRAFT, CRAFT_ICS_892/1300/1632, perfect. Default
```
[the remark this answers is not in the document]
Comment (lluism) [reply docs/scripts.rst:142]: done

File: frb/scripts/pz_dm.py [hunk 9]
```diff
@@ -1,7 +1,7 @@
-#!/usr/bin/env python
+
 """
 Estimate p(z|DM) for an assumed location on the sky and DM_FRB
-  Warning:  This assumes a *perfect* telescope
+Defaults to using the CHIME telescope model for the DM-z grid
```
[the remark this answers is not in the document]
Comment (lluism) [reply frb/scripts/pz_dm.py:4]: done

File: docs/scripts.rst [hunk 10]
```diff
@@ -107,11 +107,15 @@ This script takes as input the FRB DM and its coordinates (approximate
 are fine) and then estimates the redshift range assuming 
 the Macquart relation (and simple host + MW contributions, optionally 
 input).  For an (optionally input; tuple) confidence interval, 
-it reports back the putative redshift range for the FRB.
+it reports back the putative redshift range for the FRB. It also 
+allows for plotting the host redshift range on the magnitude vs redshift 
+evolution and setting a title for the figure.  These calculations can be 
+done assuming a few different telescope models or a perfect telescope model.
```
[the remark this answers is not in the document]
Comment (lluism) [reply docs/scripts.rst:113]: done

File: frb/tests/test_frbgalaxies.py [hunk 11]
```diff
@@ -150,3 +150,21 @@ def test_mag_dm_figure():
                flipy=True, known_hosts = False)
     assert os.path.exists('./temp_fig.png') 
     os.system('rm ./temp_fig.png')
+
+
+def test_pzdm_telescopes():
```
[the remark this answers is not in the document]
Comment (lluism) [reply frb/tests/test_frbgalaxies.py:155]: done

**Review by profxj — APPROVED** [review]
great stuff!

**Comment by lluism** [comment]
Changes:
- now the user can input MW_halo_DM and host_DM separately (both have a default value of 50). This separation is needed
   to compute DM cosmic.
- When using the "perfect" telescope, the code computes and uses DM_Cosmic instead of Extragalactic

**Review by profxj — APPROVED** [review]
all good!

**Pull request merged**